### ISSUE TITLE
C♯ completer may start OmnisharpServer twice

### ISSUE THREAD

**Issue by micbou** [Valloric/ycmd#284]
### Problem

In some occasions, OmnisharpServer will be started twice. Second instance will crash because it is using the same stdout and stderr as the first one, resulting in this on Windows:

![omnisharp-crash](https://cloud.githubusercontent.com/assets/10026824/12011069/1ebd037c-acc0-11e5-80f4-128dbb09942d.png)
*No thanks.*

Here's the log:
```
2015-12-27 17:29:13,642 - INFO - Selected solution file C:\Users\micbou\tests\ycm-issue-1744\cs\test.sln as it is the first one found
2015-12-27 17:29:13,642 - INFO - startup
2015-12-27 17:29:13,642 - INFO - Loading solution file C:\Users\micbou\tests\ycm-issue-1744\cs\test.sln
2015-12-27 17:29:13,642 - INFO - using port 1040
2015-12-27 17:29:13,644 - INFO - Received event notification
2015-12-27 17:29:13,644 - DEBUG - Event name: BufferVisit
2015-12-27 17:29:13,648 - INFO - Received event notification
2015-12-27 17:29:13,648 - DEBUG - Event name: BufferVisit
2015-12-27 17:29:13,650 - INFO - Received event notification
2015-12-27 17:29:13,651 - DEBUG - Event name: FileReadyToParse
2015-12-27 17:29:13,651 - INFO - Adding buffer identifiers for file: C:\Users\micbou\tests\ycm-issue-1744\cs\test.cs
2015-12-27 17:29:13,653 - INFO - Starting OmniSharp server
2015-12-27 17:29:13,658 - INFO - Starting new HTTP connection (1): localhost
2015-12-27 17:29:13,861 - INFO - startup
2015-12-27 17:29:13,861 - INFO - Loading solution file C:\Users\micbou\tests\ycm-issue-1744\cs\test.sln
2015-12-27 17:29:13,861 - INFO - using port 1040
2015-12-27 17:29:13,865 - INFO - Starting OmniSharp server
```
We can see that the server is started twice with same solution and same port.

### Solution

Using a lock when starting (and stopping) the server like the Javascript completer.

**Comment by micbou** [comment]
Fixed by PR #310.

**Comment by micbou** [comment]
I spoke too soon. PR #310 fixes Valloric/YouCompleteMe#1913 but this issue can still happen. Looks like using some kind of lock is necessary.

**Comment by mispencer** [comment]
I don't know if it's what you're seeing, but start server doesn't check if there is a server already running, so you could get double servers that way.

**Comment by micbou** [comment]
We are already checking it [here](https://github.com/Valloric/ycmd/blob/master/ycmd/completers/cs/cs_completer.py#L233). We could move the check directly in `StartServer` but it would still not prevent two threads to start the server simultaneously. We need a lock.

**Comment by mispencer** [comment]
I was thinking of the case where you run the StartServer completer command directly

**Comment by micbou** [comment]
Yes, this is also an issue. I'll send a PR to solve both issues.

**Comment by micbou** [comment]
Finally fixed by PR #315.
